tests/drivers/uart/uart_baudrate_test/src/main.c

```diff
@@ -0,0 +1,280 @@
+
+static const struct gpio_dt_spec gpio_spec =
+	GPIO_DT_SPEC_GET_BY_IDX(DT_PATH(zephyr_user), gpios, 0);
+static const struct device *const uart_dev = DEVICE_DT_GET(DT_NODELABEL(dut));
+
+static const uint8_t tx_buf[1] = {0x00};
+
+#define PIN_STATE_SIZE 4096
+static int pin_state[PIN_STATE_SIZE] = {};
+
+#define REPEAT_NUMBER 10
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+static void uart_fifo_callback(const struct device *dev, void *user_data)
+{
+	int ret;
+
+	ARG_UNUSED(user_data);
+
+	ret = uart_irq_update(dev);
+	zassert_true(ret >= 0, "uart_irq_update: %d\n", ret);
+
+	ret = uart_irq_tx_ready(dev);
+	zassert_true(ret >= 0, "uart_irq_tx_ready: %d\n", ret);
+
+	ret = uart_fifo_fill(dev, (uint8_t *)&tx_buf, sizeof(tx_buf));
+	zassert_true(ret == sizeof(tx_buf), "uart_fifo_fill: %d\n", ret);
+	uart_irq_tx_disable(dev);
+}
+#endif
+
+static void check_timing(const struct gpio_dt_spec *gpio_dt, uint32_t baudrate)
+{
+	uint64_t cycles_s_sys;
+	struct uart_config test_uart_config;
+	int ret;
+	uint32_t cycle_start_time;
+	uint32_t cycle_stop_time;
+	double gpio_read_time_us;
+	int32_t start_index;
+	int32_t stop_index;
+	int32_t start_index_count_zero;
+	double bit_diviation_mean;
+	double symbol_diviation_mean;
+
+	ret = uart_config_get(uart_dev, &test_uart_config);
+	zassert_equal(ret, 0, "uart_config_get: %d\n", ret);
+
+	test_uart_config.parity = UART_CFG_PARITY_EVEN;
+	test_uart_config.stop_bits = UART_CFG_STOP_BITS_2;
+	test_uart_config.flow_ctrl = UART_CFG_FLOW_CTRL_NONE;
+	test_uart_config.baudrate = baudrate;
+	ret = uart_configure(uart_dev, &test_uart_config);
+	if (ret == -ENOTSUP) {
+		TC_PRINT("[%d] Not supported\n", baudrate);
+		ztest_test_skip();
+	}
+	zassert_equal(ret, 0, "uart_configure: %d\n", ret);
+
+	cycles_s_sys = (uint64_t)sys_clock_hw_cycles_per_sec();
+	TC_PRINT("Cycles: %llu cycles\n", cycles_s_sys);
+
+	/*
+	 * Measure time needed to read gpios
+	 */
+	cycle_start_time = k_cycle_get_32();
+	for (uint32_t i = 0; i < PIN_STATE_SIZE; ++i) {
+		pin_state[i] = gpio_pin_get_dt(&gpio_spec);
+	}
+	cycle_stop_time = k_cycle_get_32();
+	gpio_read_time_us = ((cycle_stop_time - cycle_start_time) / (double)PIN_STATE_SIZE) *
+			    (1e6 / cycles_s_sys);
+	TC_PRINT("GPIO get takes: %.2f us\n", gpio_read_time_us);
+
+	start_index_count_zero = 0;
+	bit_diviation_mean = 0;
+	symbol_diviation_mean = 0;
+	for (uint32_t t = 0; t < REPEAT_NUMBER; ++t) {
+		/*
+		 * Send character
+		 */
+
+#ifdef CONFIG_UART_INTERRUPT_DRIVEN
+		uart_irq_tx_enable(uart_dev);
+#else
+		uart_tx(uart_dev, tx_buf, sizeof(tx_buf), 0);
+#endif
+
+		/*
+		 * Check gpio
+		 */
+		for (uint32_t i = 0; i < PIN_STATE_SIZE; ++i) {
+			pin_state[i] = gpio_pin_get_dt(&gpio_spec);
+		}
+
+		/*
+		 * Find start of start bit and end of stop bit
+		 */
+		start_index = -1;
+		stop_index = -1;
+		for (uint32_t i = 0; i < PIN_STATE_SIZE; ++i) {
+			if (-1 == start_index) {
+				if (0 == pin_state[i]) {
+					start_index = i;
+				}
+			}
+			if (-1 != start_index) {
```

Comment:
```suggestion
			else {
```
Even when https://github.com/nrfconnect/sdk-nrf/pull/20178/files#diff-9a3efba17f0e7baa12416886f1e02d9a62791b6c8dc9920d87b18c390583c5faR118, the `pin_state[i]` still equals `0` so the `stop_index = i;` instruction won't be executed at the same step of the loop.

Reply:
Changed, thanks :)